Publish releases on GitHub, push translatable strings to Transifex

floscher opened the pull request:
This should more or less fix #8 and fix #12.
It might need some tweaking here and there, but normally it should work.

When Travis CI is building a tag, the build artifact is published as GitHub release.
When a commit on master branch is built, the translatable strings are sent to Transifex for translation.

As mentioned in #12, you'll need to add your [Transifex API token](https://www.transifex.com/user/settings/api/) to Travis CI (either encrypted in the `.travis.yml`, or in the [Travis settings](https://travis-ci.org/osmlab/josm-atlas/settings)).

floscher committed:
Go through build.gradle, update and reformat

* put version numbers in the beginning to ease updating, drop dependencies.gradle
* reactivate `javadoc` task, seems to work now
* set up signing of artifacts (binary, sources and JavaDoc)

floscher committed:
Set up build stages for Travis CI, deploy tags to GitHub

The build stages make the `*-gate.sh` scripts obsolete (later stages only execute if the previous one succeeded).

floscher committed:
Add Transifex configuration

When building commits from the master branch on Travis CI, the translated strings are being sent to Transifex for translation: https://www.transifex.com/josm/josm/

matthieun commented on build.gradle:
```diff
@@ -14,48 +14,59 @@ plugins {
     id 'idea'
 }
 
+// Define tool and dependency versions
```
Any reason this is here and the `dependencies.gradle` is removed?

floscher replied:
I found it a bit confusing at first to have the dependency block in `build.gradle` and the actual definition of the dependency in the `dependencies.gradle`. My priority was here to have all the version numbers all in a common place for easier updating.
As long as there is only one dependency, I think it's not really needed to spread the build script over multiple files.
In the end it boils down to personal preference. If you'd like, we could keep the `dependencies.gradle`, but then maybe also move the `dependencies {}` block over there and as much of the version numbers as easily possible.

matthieun commented on build.gradle:
```diff
@@ -66,54 +77,42 @@ def skippedTaskNames = System.getenv().findAll { key, value ->
+    ext."signing.password" = System.env.GPG_PASSPHRASE
+    ext."signing.secretKeyRingFile" = System.env.GPG_KEY_LOCATION
+}
+if (!ext.has("signing.keyId") || !ext.has("signing.password") || !ext.has("signing.secretKeyRingFile")) {
```
Is it possible to fail the build in that case? It would be better to have the artifacts signed always.

floscher replied:
It's possible, but builds that are not running on Travis should still be possible without signing setup. I'll add a condition for that.

floscher replied:
I added 0c58af8d0116f611b1b26ef235ff7ae0bb0ce808 below to address this.
The build now fails if all of the following conditions are met:
* the environment variable `TRAVIS_REPO_SLUG` is set to `osmlab/josm-atlas`
* the environment variable `TRAVIS_PULL_REQUEST` is set to `false` (PRs can't access encrypted variables for security reasons)
* the signing options are not set up

matthieun commented on .travis.yml:
```diff
@@ -12,17 +12,66 @@ env:
+stages:
+  - name: test
+  - name: i18n
+    if: branch = master AND type = push AND repo = osmlab/josm-atlas
```
Thanks, this is way cleaner than using scripts!

matthieun reviewed (commented):
Thanks, looks great! A few questions...

hallahan commented:
Looks good. LGTM!

floscher committed:
Fail build in CI when no signing information is given

When building the repo `oamlab/josm-atlas` on Travis CI, fail the build when signing of the release jar is not set up.

hallahan commented:
What is the status of this PR?

floscher commented:
From my perspective it is ready. If the `dependencies.gradle` should not be removed, I could exclude that part from the PR. Or if there's any other part I should explain more or change in some way, feel free to ask.

What still needs to be done if this PR gets merged:
- [ ] I'd have to give a Transifex user the necessary privileges to upload to https://transifex.com/josm/josm
- [ ] That user would have to add [his/her Transifex API](https://www.transifex.com/user/settings/api/) token to Travis CI as environment variable `TRANSIFEX_TOKEN`

hallahan commented:
Looks Great! Thank you for your valuable contribution. My apologies for taking so long to merge this.

floscher commented:
Great, thanks. Should I add [your Transifex account](https://www.transifex.com/user/profile/hallahan/) to the JOSM team, so your API can be used for uploading to Transifex?
